chat_ui/src/main/kotlin/io.skygear.plugins.chat.ui/ConversationFragment.kt

```diff
@@ -530,34 +570,64 @@ open class ConversationFragment() :
             val message = ChatMessage()
             message.asset = asset
             message.metadata = meta
-            addMessageToBottom(message, Uri.parse("file://" + voiceRecordingFileName))
 
+            this.fragmentMessageSentListener?.onBeforeMessageSent(this, message)
+            addMessageToBottom(message, Uri.parse("file://" + voiceRecordingFileName))
             this.skygearChat?.addMessage(message, conv, object : SaveCallback<ChatMessage> {
                 override fun onSuccess(chatMsg: ChatMessage?) {
                     voiceRecordingFile.delete()
+                    callMessageSentSuccessListener(message)
                 }
 
                 override fun onFail(error: Error) {
                     Log.e(
                             ConversationFragment.TAG,
                             "Failed to send voice message: ${error.message}"
                     )
+                    this@ConversationFragment.fragmentMessageSentListener?.onMessageSentFailed(this@ConversationFragment, message, error)
                 }
             })
         }
     }
 
+    fun callMessageSentSuccessListener(chatMsg: ChatMessage?) {
+        chatMsg?.let {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentSuccess(this@ConversationFragment, chatMsg)
+        } ?: run {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentFailed(this@ConversationFragment,
+                            null, Error("No message sent"))
```

Comment:
It is not called in practice. However, since SaveCallback's void onSuccess(@Nullable T object) is nullable, and I should not call onMessageSentSuccess if message is null, right?